### PR TITLE
add error for invalid friend compare code

### LINKED ISSUE #6: Display error message for invalid friend codes
https://github.com/liftM/hipsterfy/blob/2af2a3c939b4f29aac0cdb8adca513d7300b5c21/cmd/hipsterfy/Main.hs#L168-L174

### PR DESCRIPTION
**Pull request by emilyzhang**
- adds error when friend compare code is invalid
- fixes https://github.com/liftM/hipsterfy/issues/6
- screenshot:
image

### CODE REVIEW

File: cmd/hipsterfy/Hipsterfy/Server.hs [hunk 1]
```diff
@@ -74,20 +74,19 @@ handleCompare = post "/compare" $ do
...
 
-  -- Render page.
-  yourStatus <- infoToStatus $ updateJobInfo user
-  friendStatus <- infoToStatus $ updateJobInfo friend
-  html $ comparePage (yourStatus, yourArtists) (friendStatus, friendArtists)
+      -- Render page.
+      yourStatus <- infoToStatus $ updateJobInfo user
+      friendStatus <- infoToStatus $ updateJobInfo f
+      html $ comparePage $ Right ((yourStatus, yourArtists), (friendStatus, friendArtists))
+    Nothing -> do html $ comparePage $ Left FriendCodeError
```
Comment (liftM) [inline cmd/hipsterfy/Hipsterfy/Server.hs:88]: This should really just be its own page. It has literally nothing in common with the page rendered in the `Right` path.

File: cmd/hipsterfy/Hipsterfy/Server/Pages.hs [hunk 2]
```diff
@@ -14,6 +13,8 @@ import Text.Blaze.Html5 hiding (body, contents, head, map)
 import qualified Text.Blaze.Html5 as H
 import qualified Text.Blaze.Html5.Attributes as A
 
+data PageError = FriendCodeError
```
Comment (liftM) [inline cmd/hipsterfy/Hipsterfy/Server/Pages.hs:16]: This type isn't very useful - it doesn't actually carry any information, isn't part of a sum, and if you _really_ wanted to use `Either`, you could use either `Either () r` or its isomorphic equivalent `Maybe r`.

**Review by elldritch — CHANGES_REQUESTED** [review]

File: cmd/hipsterfy/Hipsterfy/Server.hs [hunk 3]
```diff
@@ -74,20 +74,19 @@ handleCompare = post "/compare" $ do
...
 
-  -- Render page.
-  yourStatus <- infoToStatus $ updateJobInfo user
-  friendStatus <- infoToStatus $ updateJobInfo friend
-  html $ comparePage (yourStatus, yourArtists) (friendStatus, friendArtists)
+      -- Render page.
+      yourStatus <- infoToStatus $ updateJobInfo user
+      friendStatus <- infoToStatus $ updateJobInfo f
+      html $ comparePage (yourStatus, yourArtists) (friendStatus, friendArtists)
+    Nothing -> do html $ friendErrorPage
```
Comment (liftM) [inline cmd/hipsterfy/Hipsterfy/Server.hs:88]: The `do` here is redundant.

File: cmd/hipsterfy/Hipsterfy/Server/Pages.hs [hunk 4]
```diff
@@ -98,6 +97,9 @@ comparePage (x, xs) (y, ys) = container $ do
   br
   href "/" "Go back"
 
+friendErrorPage :: LText
+friendErrorPage = container $ do div ! A.class_ "self-center mt-5 text-red-700" $ "Error: This is an invalid friend code."
```
Comment (liftM) [inline cmd/hipsterfy/Hipsterfy/Server/Pages.hs:101]: The `do` here is also redundant.

File: cmd/hipsterfy/Hipsterfy/Server.hs [hunk 5]
```diff
@@ -74,20 +74,19 @@ handleCompare = post "/compare" $ do
...
-    Just f -> return f
-    -- TODO: display an error message for invalid friend codes.
-    Nothing -> redirect "/"
-
-  -- Load followed artists.
-  -- TODO: load these in parallel
-  yourArtists <- getFollowedArtists $ userID user
-  friendArtists <- getFollowedArtists $ userID friend
+  case maybeFriend of
+    Just f -> do
```
Comment (liftM) [inline cmd/hipsterfy/Hipsterfy/Server.hs:78]: [nit] Since `friend` is no longer a used name, I'd expand `f` to `friend` for readability.

**Review by elldritch — CHANGES_REQUESTED** [review]
LGTM after nits. Can you also attach a screenshot of how this page looks?

**Review by elldritch — APPROVED** [review]
LGTM.

**Pull request merged**